packages/core/src/defer/triggering.ts

```diff
@@ -337,61 +337,135 @@ export function triggerDeferBlock(lView: LView, tNode: TNode) {
+  // Trigger resource loading and hydration for the blocks in the queue in the order of highest block
+  // to lowest block. Once a block has finished resource loading, after next render, and onHydrationComplete
+  // has fired, the new block will have had its defer instruction called and will be in the registry.
+  // Due to timing related to potential nested control flow, this has to be scheduled after the next render.
+
+  // Indicate that we have some pending async work.
+  const pendingTasks = injector.get(PendingTasksInternal);
+  const taskId = pendingTasks.add();
+
+  // If the parent block was being hydrated, but the process was
+  // not yet complete, wait until parent block promise settles before
+  if (parentBlockPromise !== null) {
+    await parentBlockPromise;
+  }
+
+  // Actually do the triggering and hydration of the queue of blocks
+  for (const dehydratedBlockId of hydrationQueue) {
+    await nextRender(injector);
+    await triggerDeferBlockResourceLoading(dehydratedBlockId, dehydratedBlockRegistry);
+    // TODO(incremental-hydration): call `rejectFn` when lDetails[DEFER_BLOCK_STATE] is `DeferBlockState.Error`.
+    blocksBeingHydrated.get(dehydratedBlockId)!.resolveFn()();
+  }
+
+  // Await hydration completion for the requested block.
+  await blocksBeingHydrated.get(blockName)?.promise;
+
+  // Replay is timing dependent and needs to happen after the next render.
+  await nextRender(injector);
+
+  // All async work is done, remove the taskId from the registry.
+  pendingTasks.remove(taskId);
+
+  // Phase 3: Replay any queued events, if any exist
+  if (replayQueuedEventsFn) {
+    replayQueuedEventsFn(hydrationQueue);
+  }
+
+  // Phase 4: Cleanup
+  cleanupHydratedDeferBlocks(
+    dehydratedBlockRegistry.get(blockName),
+    hydrationQueue,
+    dehydratedBlockRegistry,
+    injector.get(ApplicationRef),
+  );
 }
 
 /**
- * Triggers the resource loading for a defer block and passes back a promise
- * to handle cleanup on completion
+ * Generates a new promise for every defer block in the hydrating queue
  */
-export function triggerAndWaitForCompletion(
-  dehydratedBlockId: string,
-  dehydratedBlockRegistry: DehydratedBlockRegistry,
-  injector: Injector,
-): Promise<void> {
-  // TODO(incremental-hydration): This is a temporary fix to resolve control flow
-  // cases where nested defer blocks are inside control flow. We wait for each nested
-  // defer block to load and render before triggering the next one in a sequence. This is
-  // needed to ensure that corresponding LViews & LContainers are available for a block
-  // before we trigger it. We need to investigate how to get rid of the `afterNextRender`
-  // calls (in the nearest future) and do loading of all dependencies of nested defer blocks
-  // in parallel (later).
+function populateHydratingStateForQueue(registry: DehydratedBlockRegistry, queue: string[]) {
+  for (let blockId of queue) {
+    registry.hydrating.set(blockId, getPromise());
```

Comment:
It looks like we can just use a standard `Promise.withResolvers()` here:
```suggestion
    registry.hydrating.set(blockId, Promise.withResolvers());
```

Reply:
We can also type the `hydrating` Map accordingly - either using a standard typing or a custom type that we can then reuse in defer code.

Reply:
Hmmm, it would seem typescript doesn't support `withResolvers()` yet, which is surprising.

Reply:
![Screenshot 2024-11-21 at 10 04 36 AM](https://github.com/user-attachments/assets/02bb39c0-60c1-4baa-aa0f-17d8ac837026)